### PR TITLE
Story/cord 33

### PR DESCRIPTION
**Pull request by aa-kashk**
There should be a browse option to browse through all papers

### CODE REVIEW

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/PublicationDao.java [hunk 1]
```diff
@@ -40,4 +41,6 @@
 
     long getTotalAffiliation();
 
+    List<PublicationImpl> getPublicationTitles(Long offset, Integer size);
```
Comment (jdamerow) [inline cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/PublicationDao.java]: this makes it sound like it retrieves all titles, not all publications

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java [hunk 2]
```diff
@@ -64,6 +64,15 @@
             i++;
         }
         return results;
+    }
+    
+    @Override
+    public List<PublicationImpl> getPublicationTitles(Long offset, Integer size){
+        Criteria titles = Criteria.where("metadata.title").ne(null);
```
Comment (jdamerow) [inline cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java]: we want all publications, not just the ones with titles

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/ShowPublicationController.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+@Controller
+public class ShowPublicationController {
+
+    @Autowired
+    private PublicationDao pubDao;
+
+    @RequestMapping("/publications")
+    public String get(Model model, Pageable pageable) {
+        model.addAttribute("pubs", pubDao.getPublications(pageable));
+        long pubCount = pubDao.getPublicationCount();
```
Comment (jdamerow) [inline cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/ShowPublicationController.java]: doesn't the page that the getPublication method returns have the total number of elements?

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/ShowPublicationController.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+@Controller
+public class ShowPublicationController {
+
+    @Autowired
+    private PublicationDao pubDao;
+
+    @RequestMapping("/publications")
+    public String get(Model model, Pageable pageable) {
+        Page<PublicationImpl> pages = pubDao.getPublications(pageable);
```
Comment (jdamerow) [inline cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/ShowPublicationController.java]: should be "page" since it's just one page

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
from the ticket: "sorted by title or journal or year; this should be available on the public page"

I don't think this is done?

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/config/CollectionsConfig.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.DependsOn;
+import org.springframework.data.domain.Sort;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import org.springframework.data.mongodb.core.index.Index;
+
+import edu.asu.diging.cord19.explorer.core.model.impl.PublicationImpl;
+
+@Configuration
+@DependsOn("mongoTemplate")
```
Comment (jdamerow) [inline cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/config/CollectionsConfig.java]: do you really need this?

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java [hunk 6]
```diff
@@ -66,11 +69,22 @@
         return results;
     }
 
+    @Override
+    public Page<PublicationImpl> getPublications(Pageable pageable) {
+        Query query = new Query().with(pageable).with(Sort.by(Order.asc("metadata.title")));
```
Comment (jdamerow) [inline cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java]: sorting should come from the user: "sorted by title or journal or year"

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/ShowPublicationController.java (hunk 4 above)
Comment (jdamerow) [reply cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/ShowPublicationController.java]: ?

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/ShowPublicationController.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+import edu.asu.diging.cord19.explorer.core.model.impl.PublicationImpl;
+import edu.asu.diging.cord19.explorer.core.mongo.PublicationDao;
+
+@Controller
+public class ShowPublicationController {
+
+    @Autowired
+    private PublicationDao pubDao;
+
+    @RequestMapping("/pubs")
```
Comment (jdamerow) [inline cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/ShowPublicationController.java]: why is that called `pubs` and not `publications`?

File: cord19-explorer/src/main/webapp/WEB-INF/views/publications.html [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+		    <li th:if="${page + 1 < pageCount}"><a href="#" th:href="@{/pubs(page=${page+1}, sort=${sort})}">[[${page+2}]]</a></li>
+		    <li th:if="${page + 2 < pageCount}"><a href="#" th:href="@{/pubs(page=${page+2}, sort=${sort})}">[[${page+3}]]</a></li>
+		    <li th:if="${page + 3 < pageCount}"><a href="#" th:href="@{/pubs(page=${page+3}, sort=${sort})}">[[${page+4}]]</a></li>
+		    <li th:if="${page + 4 < pageCount}"><a href="#" th:href="@{/pubs(page=${page+4}, sort=${sort})}">[[${page+5}]]</a></li>
+		    
+		    <li th:if="${page + 1 < pageCount}"><a href="#" th:href="@{/pubs(page=${page+1}, sort=${sort})}" aria-label="Previous"><span aria-hidden="true">&raquo;</span></a></li>
+		    <li th:if="${page + 1 >= pageCount}" class="disabled"><a href="#" aria-label="Previous"><span aria-hidden="true">&raquo;</span></a></li>
+		    
+		    <li th:if="${page != pageCount-1}"><a href="#" th:href="@{/pubs(page=${pageCount-1}, sort=${sort})}" aria-label="Last"><span aria-hidden="true">&raquo;&raquo;</span></a></li>
+		  	<li th:if="${page == pageCount-1}" class="disabled"><a href="#" aria-label="Last"><span aria-hidden="true">&raquo;&raquo;</span></a></li>
```
Comment (jdamerow) [inline cord19-explorer/src/main/webapp/WEB-INF/views/publications.html]: indentation (also further up)

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Review by jdamerow — CHANGES_REQUESTED** [review]
- by default it the list seems to be sorted by title desc, which doesn't make sense. It should start with a and end with z, not the other way around ;)
- year should be shown for publications
- journal should be shown for publications 
(if these two are not shown, the sorting is kind of useless to an extend)
- it needs to be possible to change from descending to ascending order
- there needs to be a blank between "Sort by: " and the dropdown

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/ShowPublicationController.java (hunk 7 above)
Comment (aa-kashk) [reply cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/ShowPublicationController.java]: `/publications` is used by `PublicAffiliationController` in merged Story 26.

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java [hunk 9]
```diff
@@ -67,20 +73,35 @@
...
+        String[] sort = pageable.getSort().toString().split(": ");
+
+        long startItem = pageable.getPageNumber() * pageable.getPageSize();
+        int limit = pageable.getPageSize();
+        List<PublicationImpl> list = StreamUtils.createStreamFromIterator(mongoTemplate.stream(
+                new Query().with(Sort.by(new Order(sort[1].equals("ASC") ? Direction.ASC : Direction.DESC, sort[0]))),
+                PublicationImpl.class)).skip(startItem).limit(limit).collect(Collectors.toList());
+
+        Page<PublicationImpl> page = new PageImpl<PublicationImpl>(list, pageable, list.size());
+        return page;
```
Comment (jdamerow) [inline cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java]: merge with line 87

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/ShowPublicationController.java (hunk 7 above)
Comment (jdamerow) [reply cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/ShowPublicationController.java]: let's rename the other one to `/affiliation/publications` or something and make this one `/publications`

File: cord19-explorer/src/main/webapp/WEB-INF/views/publications.html (hunk 8 above)
Comment (jdamerow) [reply cord19-explorer/src/main/webapp/WEB-INF/views/publications.html]: ?

File: cord19-explorer/src/main/webapp/WEB-INF/views/layouts/navbar.html [hunk 10]
```diff
@@ -39,6 +39,8 @@ <h1>Ups, something went wrong.</h1>
    <div id="mySidenav" class="sidenav">
 	  <button class="btn dark-btn" onclick="closeNav()">&times;</button>
 	  <ul class="nav nav-pills">
+	  	  <li role="presentation">
+	         <a th:href="@{/pubs(sort='metadata.title')}" >Publications</a>
```
Comment (jdamerow) [inline cord19-explorer/src/main/webapp/WEB-INF/views/layouts/navbar.html]: indentation and missing `</li>`

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
Unfortunately, using the dbcursor, didn't fix the performance issue (you'll notice if you page through the review instance, and I assume you would notice locally as well). After my above comments are fixed, I'll merge this story, but we need to find a better solution. I'll think about it, and create another story.

**Comment by jdamerow** [comment]
Actually, I changed my mind. Can we try to query using the `$gt` parameter? Something like: `{"metadata.title": { $gt: "WHO LAUNCHES NEW GLOBAL INFLUENZA STRATEGY"}}` combined with sort and limit. Let's see if that changes anything.

**Comment by jdamerow** [comment]
actually, it probably needs to be `gte` and `gt` the last id since titles can appear multiple time. This also means, you would have to sort by first title, and then id. For the next page you would say equal to the last title and id is greater than the last id, or greater than the last title.

**Comment by jdamerow** [comment]
Should probably look something like this
`{ $or: [{"metadata.title": {$gt : <previous title>}}, {"metadata.title": {$eq: <previous title>}, "_id": {$gt: <last id>}}]}`

**Comment by aa-kashk** [comment]
So, when I include a criteria for this ` "_id": {$gt: <last id>}`, the result is empty. When I include the Sort operation in the query based on title and id, it works for the first few pages but it throws 'Sort operation used more than the maximum 33554432 bytes of RAM' error.

**Comment by jdamerow** [comment]
Make it so, Jenkins.

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java [hunk 11]
```diff
@@ -68,47 +72,83 @@
         return results;
     }
 
+    @Override
+    public List<PublicationImpl> getPublications(Pageable pageable, PublicationImpl pub, boolean init) {
```
Comment (jdamerow) [inline cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java:76]: what is `init` for? also, javadoc

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/model/SortType.java [hunk 12]
```diff
@@ -0,0 +1,5 @@
+package edu.asu.diging.cord19.explorer.web.model;
+
+public enum SortType {
+    publishYear, journal
```
Comment (jdamerow) [inline cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/model/SortType.java:4]: please follow our style guide

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java [hunk 13]
```diff
@@ -68,47 +72,83 @@
...
+
+        String[] title = pageable.getSort().toString().split(": ");
```
Comment (jdamerow) [inline cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java:78]: why is this variable called `title`?

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java [hunk 14]
```diff
@@ -68,47 +72,83 @@
...
+            criteria2 = criteria2.andOperator(Criteria.where("id").gt(new ObjectId(pub.getId().toString())));
+
+        } else {
+            if (title[0].equals(SortType.publishYear.toString())) {
+                criteria1 = Criteria.where(title[0]).gt(0);
+                criteria2 = Criteria.where(title[0]).is(0);
+            } else {
+                criteria1 = Criteria.where(title[0]).gt("");
+                criteria2 = Criteria.where(title[0]).is("");
+            }
```
Comment (jdamerow) [inline cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java:103]: what about journal in this else block?

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java [hunk 15]
```diff
@@ -68,47 +72,83 @@
...
+            if (title[0].equals(SortType.publishYear.toString())) {
+                criteria1 = Criteria.where(title[0]).gt(pub.getPublishYear());
+                criteria2 = Criteria.where(title[0]).is(pub.getPublishYear());
+            } else if (title[0].equals(SortType.journal.toString())) {
+                criteria1 = Criteria.where(title[0]).gt(pub.getJournal());
+                criteria2 = Criteria.where(title[0]).is(pub.getJournal());
+            } else {
+                criteria1 = Criteria.where(title[0]).gt(pub.getMetadata().getTitle());
+                criteria2 = Criteria.where(title[0]).is(pub.getMetadata().getTitle());
+            }
```
Comment (jdamerow) [inline cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java:92]: these huge if/else blocks are not very maintainable. What if we want to add more options for sorting? I would try to maybe just have a map that maps fields to methods that are then called.

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java (hunk 14 above)
Comment (jdamerow) [reply cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java:103]: DRY (don't repeat yourself)

**Review by jdamerow — CHANGES_REQUESTED** [review]
There is a bug somehwere. If I go from page one to page two and back to page one, the results I see are different than before.

**Comment by jdamerow** [comment]
Also, if you haven't done so yet, please read the pull request guidelines in confluence.

**Comment by jdamerow** [comment]
also #2, please write better commit messages. "backend 4" is not a good message, what does that mean?

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java (hunk 11 above)
Comment (aa-kashk) [reply cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java:76]: I might be getting rid of the `init` variable in the new implementation. Added a javadoc for this method.

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/model/SortType.java (hunk 12 above)
Comment (aa-kashk) [reply cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/web/model/SortType.java:4]: I checked the enums in Monitor and they are defined in capital letters.

File: cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java (hunk 13 above)
Comment (aa-kashk) [reply cord19-explorer/src/main/java/edu/asu/diging/cord19/explorer/core/mongo/impl/PublicationDaoImpl.java:78]: changed this to `sortType`

**Comment by aa-kashk** [comment]
I tried using the $lt operator on the first id of the page 2 and higher to return a list for the previous page but it doesn't seem to fix the issue.

**Pull request closed without merging**